flink-runtime/src/test/java/org/apache/flink/runtime/heartbeat/TestingHeartbeatServices.java

```diff
@@ -18,12 +18,175 @@
+		super(heartbeatInterval, heartbeatTimeout);
+	}
+
+	@Override
+	public <I, O> HeartbeatManager<I, O> createHeartbeatManager(
+		ResourceID resourceId,
+		HeartbeatListener<I, O> heartbeatListener,
+		ScheduledExecutor mainThreadExecutor,
+		Logger log) {
+
+		HeartbeatManagerImpl<I, O> heartbeatManager = new HeartbeatManagerImpl<>(
+			heartbeatTimeout,
+			resourceId,
+			heartbeatListener,
+			mainThreadExecutor,
+			log,
+			new TestingHeartbeatMonitorFactory<>());
+
+		heartbeatManagers.put(resourceId, heartbeatManager);
+		return heartbeatManager;
+	}
+
+	@Override
+	public <I, O> HeartbeatManager<I, O> createHeartbeatManagerSender(
+		ResourceID resourceId,
+		HeartbeatListener<I, O> heartbeatListener,
+		ScheduledExecutor mainThreadExecutor,
+		Logger log) {
+
+		HeartbeatManagerSenderImpl<I, O> heartbeatManager = new HeartbeatManagerSenderImpl<>(
+			heartbeatInterval,
+			heartbeatTimeout,
+			resourceId,
+			heartbeatListener,
+			mainThreadExecutor,
+			log,
+			new TestingHeartbeatMonitorFactory<>());
+
+		heartbeatManagerSenders.put(resourceId, heartbeatManager);
+		return heartbeatManager;
+	}
+
+	public void triggerHeartbeatTimeout(ResourceID managerResourceId, ResourceID targetResourceId) {
+
+		boolean triggered = false;
+		HeartbeatManagerImpl heartbeatManager = heartbeatManagers.get(managerResourceId);
+		if (heartbeatManager != null) {
+			final TestingHeartbeatMonitor monitor =
+				(TestingHeartbeatMonitor) heartbeatManager.getHeartbeatTargets().get(targetResourceId);
+			if (monitor != null) {
+				monitor.triggerHeartbeatTimeout();
+				triggered = true;
+			}
+		}
+
+		heartbeatManager = heartbeatManagerSenders.get(managerResourceId);
+		if (heartbeatManager != null) {
+			final TestingHeartbeatMonitor monitor =
+				(TestingHeartbeatMonitor) heartbeatManager.getHeartbeatTargets().get(targetResourceId);
+			if (monitor != null) {
+				monitor.triggerHeartbeatTimeout();
+				triggered = true;
+			}
+		}
+
+		checkState(triggered,
+			"There is no target " + targetResourceId + " monitored under Heartbeat manager " + managerResourceId);
+	}
+
+	/**
+	 * Factory instantiates testing monitor instance.
+	 *
+	 * @param <O> Type of the outgoing heartbeat payload
+	 */
+	static class TestingHeartbeatMonitorFactory<O> implements HeartbeatMonitor.Factory<O> {
+
+		@Override
+		public HeartbeatMonitor<O> createHeartbeatMonitor(
+			ResourceID resourceID,
+			HeartbeatTarget<O> heartbeatTarget,
+			ScheduledExecutor mainThreadExecutor,
+			HeartbeatListener<?, O> heartbeatListener,
+			long heartbeatTimeoutIntervalMs) {
+
+			return new TestingHeartbeatMonitor<>(
+				resourceID,
+				heartbeatTarget,
+				mainThreadExecutor,
+				heartbeatListener,
+				heartbeatTimeoutIntervalMs);
+		}
+	}
+
+	/**
+	 * A heartbeat monitor for testing which supports triggering timeout manually.
+	 *
+	 * @param <O> Type of the outgoing heartbeat payload
+	 */
+	static class TestingHeartbeatMonitor<O> extends HeartbeatMonitorImpl<O> {
```

Comment:
Couldn't we also implement a very simple `HeartbeatMonitor` which keeps a reference to the `ScheduledExecutor` and the `HeartbeatListener` and then simply calls `scheduledExecutor.execute(() -> heartbeatListener.notifyHeartbeatTimeout)` if `triggerHeartbeatTimeout` is called? The current implementation relies a lot on the underlying implementation details of `HeartbeatMonitorImpl`.

Reply:
Yes you are right, it indeed relies on a bit more implementation of `HeartbeatMonitorImpl`.
I think your suggestion could work well.

I'd like to make some details clear.

I think we should keep the heartbeat working normally before triggering a timeout manually. Otherwise we might miss some important information like an unexpected heartbeat timeout.